common/djangoapps/student/models.py

```diff
@@ -1002,7 +1005,6 @@ def __unicode__(self):
         ).format(self.user, self.course_id, self.created, self.is_active)
 
     @classmethod
-    @transaction.atomic
```

Comment:
Potentially, you could look at shrinking the transaction to more precisely surround the database changes that @rlucioni mentions, without surrounding things like request parsing, validation, and result serialization. I'm not sure if that would give you a sufficiently short transaction to reduce the number of integrity errors, though.